### PR TITLE
Feat(hints) 55, 56 & 57

### PR DESCRIPTION
**Pull request by jpcenteno**
# Feat(hints) 55, 56 & 57

### Implement `NewHint#55` from `vrf.json`

```
from starkware.cairo.common.cairo_secp.secp_utils import pack
SECP_P=2**255-19

x = pack(ids.x, PRIME) % SECP_P

```

### Implement `NewHint#56` from `vrf.json`

```
from starkware.cairo.common.cairo_secp.secp_utils import pack
SECP_P=2**255-19

value = pack(ids.x, PRIME) % SECP_P

```

### Implement `NewHint#57` from `vrf.json`

```
SECP_P=2**255-19
from starkware.python.math_utils import div_mod

value = x_inv = div_mod(1, x, SECP_P)

```
Closes: #979

## Checklist
- [ ] Linked to Github Issue
- [ ] Unit tests added
- [ ] Integration tests added.
- [ ] This change requires new documentation.
  - [x] Documentation has been added/updated.
  - [ ] CHANGELOG has been updated.

### CODE REVIEW

File: src/hint_processor/builtin_hint_processor/vrf/pack.rs [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+///
+/// x = pack(ids.x, PRIME) % SECP_P
+/// ```
+pub fn assign_pack_mod_secp_prime_to_x(
+    vm: &mut VirtualMachine,
+    exec_scopes: &mut ExecutionScopes,
+    ids_data: &HashMap<String, HintReference>,
+    ap_tracking: &ApTracking,
+) -> Result<(), HintError> {
+    let x = BigInt3::from_var_name("x", vm, ids_data, ap_tracking)?.pack86();
```
Comment (pefontana) [inline src/hint_processor/builtin_hint_processor/vrf/pack.rs:29]: Assing the `SECP` to the scope

File: src/hint_processor/builtin_hint_processor/vrf/pack.rs [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+///
+/// value = pack(ids.x, PRIME) % SECP_P
+/// ```
+pub fn assign_pack_mod_secp_prime_to_value(
+    vm: &mut VirtualMachine,
+    exec_scopes: &mut ExecutionScopes,
+    ids_data: &HashMap<String, HintReference>,
+    ap_tracking: &ApTracking,
+) -> Result<(), HintError> {
+    let x = BigInt3::from_var_name("x", vm, ids_data, ap_tracking)?.pack86();
```
Comment (pefontana) [inline src/hint_processor/builtin_hint_processor/vrf/pack.rs:48]: Assing the `SECP` to the scope

File: src/hint_processor/builtin_hint_processor/vrf/pack.rs [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+/// Implements hint:
+/// ```python
+/// SECP_P=2**255-19
+/// from starkware.python.math_utils import div_mod
+///
+/// value = x_inv = div_mod(1, x, SECP_P)
+/// ```
+pub fn assign_div_mod_1_x_secp_prime_to_x_inv_and_value(
+    exec_scopes: &mut ExecutionScopes,
+) -> Result<(), HintError> {
```
Comment (pefontana) [inline src/hint_processor/builtin_hint_processor/vrf/pack.rs:63]: Assing the `SECP` to the scope

**Review by pefontana — CHANGES_REQUESTED** [review]
Nice one!

File: src/hint_processor/builtin_hint_processor/vrf/pack.rs (hunk 1 above)
Comment (MegaRedHand) [reply src/hint_processor/builtin_hint_processor/vrf/pack.rs:29]: Done!

File: cairo_programs/reduce.cairo [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+    let (res) = reduce_ed25519(x);
+    assert res = BigInt3(
+        46626018522843898369279061, 61383716336970764285445175, 1091056094881698945030653
+    );
+
+    return ();
+}
+
+func main{range_check_ptr}() {
+    return ();
```
Comment (fmoletta) [inline cairo_programs/reduce.cairo:87]:
```suggestion
    test_reduce_ed25519():
    return ();
```

File: src/hint_processor/builtin_hint_processor/vrf/pack.rs [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+        x_d0: i128,
+        x_d1: i128,
+        x_d2: i128,
+        expected: BigInt,
+    ) {
+        let ids_data = non_continuous_ids_data![("x", 0)];
+
+        let mut vm = vm!();
+        vm.run_context.fp = 0;
+        add_segments!(vm, 3); // Alloc space for `ids.x.d0`, `ids.x.d1` and `ids.x.d2`.
```
Comment (fmoletta) [inline src/hint_processor/builtin_hint_processor/vrf/pack.rs:107]:
```suggestion
```

The `segments` macro already allocates the segment(s) it needs

File: src/hint_processor/builtin_hint_processor/vrf/pack.rs [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
```
Comment (fmoletta) [inline src/hint_processor/builtin_hint_processor/vrf/pack.rs:137]:
```suggestion
```

File: src/hint_processor/builtin_hint_processor/vrf/pack.rs [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+        );
+
+        let x_result = exec_scopes.get::<BigInt>("value");
+        assert!(x_result.is_ok());
+        assert_eq!(x_result.unwrap(), expected);
+
+        let secp_p = exec_scopes.get::<BigInt>("SECP_P");
+        assert!(secp_p.is_ok());
+        assert_eq!(secp_p.unwrap(), SECP_P_V2.clone());
+    }
```
Comment (fmoletta) [inline src/hint_processor/builtin_hint_processor/vrf/pack.rs:154]: You can use the macro `check_scope` for this

File: src/tests/cairo_run_test.rs [hunk 8]
```diff
@@ -939,3 +939,10 @@ fn cairo_run_normalize_address() {
     let program_data = include_bytes!("../../cairo_programs/normalize_address.json");
     run_program_simple_with_memory_holes(program_data.as_slice(), 110);
 }
+
+#[test]
+#[cfg_attr(target_arch = "wasm32", wasm_bindgen_test)]
+fn cairo_run_reduce() {
+    let program_data = include_bytes!("../../cairo_programs/reduce.json");
+    run_program_simple_with_memory_holes(program_data.as_slice(), 0);
```
Comment (fmoletta) [inline src/tests/cairo_run_test.rs:947]:
```suggestion
    run_program_simple(program_data.as_slice());
```

File: cairo_programs/reduce.cairo (hunk 4 above)
Comment (MegaRedHand) [reply cairo_programs/reduce.cairo:88]: Thanks for catching this one!

**Review by pefontana — APPROVED** [review]

**Review by fmoletta — APPROVED** [review]

**Pull request merged**